src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h

```diff
@@ -0,0 +1,298 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking for function and loop contracts.
+
+Author: Qinheping Hu, qinhh@amazon.com
+Author: Remi Delmas delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+/// \file
+/// Class that computes CFG information about the loop structure
+/// of a GOTO function for the purpose of dynamic frame conditions checking
+/// and loop contract instrumentation.
+
+#ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_CFG_INFO_H
+#define CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_CFG_INFO_H
+
+#include <util/message.h>
+#include <util/namespace.h>
+#include <util/std_expr.h>
+#include <util/symbol_table.h>
+
+#include <goto-programs/goto_program.h>
+
+#include "dfcc_loop_contract_mode.h"
+
+#include <map>
+#include <set>
+#include <unordered_set>
+
+class dfcc_utilst;
+class dfcc_libraryt;
+class goto_functiont;
+
+/// \brief Describes a single loop for the purpose of DFCC loop contract
+/// instrumentation.
+class dfcc_loop_infot
+{
+public:
+  dfcc_loop_infot(
+    std::size_t loop_id,
+    std::size_t cbmc_loop_id,
+    const std::set<exprt> &assigns,
+    exprt invariant,
+    exprt::operandst decreases,
+    const std::unordered_set<irep_idt> &local,
+    const std::unordered_set<irep_idt> &tracked,
+    std::set<std::size_t> inner_loops,
+    std::set<std::size_t> outer_loops,
+    symbol_exprt write_set_var,
+    symbol_exprt addr_of_write_set_var)
+    : loop_id(loop_id),
+      cbmc_loop_id(cbmc_loop_id),
+      assigns(assigns),
+      invariant(invariant),
+      decreases(decreases),
+      local(local),
+      tracked(tracked),
+      inner_loops(inner_loops),
+      outer_loops(outer_loops),
+      write_set_var(write_set_var),
+      addr_of_write_set_var(addr_of_write_set_var)
+  {
+  }
+
+  /// Prints a textual representation of the struct to \p out.
+  void output(std::ostream &out) const;
+
+  // Finds the instruction tagged as loop head and having the same loop
+  // identifier as this struct in the given program.
+  optionalt<goto_programt::targett>
+  find_head(goto_programt &goto_program) const;
+
+  // Finds the last instruction tagged as loop latch and having the same loop
+  // loop identifier as this struct in the given program.
+  optionalt<goto_programt::targett>
+  find_latch(goto_programt &goto_program) const;
```

Comment:
The `goto_program` should contain the loop corresponding to the `dfcc_loop_infot`.

Reply:
Technically it is not the the same program since it is going to undergo instrumentation, but I'll still add a comment to that effect.

The whole idea behind this tagging system is to use tags to locate head and latches instead of keeping references or iterators that will be invalidated by the insertion of new instructions during instrumentation.

The tags stick to the instructions and move with them, so can add instructions, transform nested loops, and still being able to locate the heads and latches using linear scans in the instruction sequence: first head tag from the start is the head, last latch tag from the end is the latch.